packages/e2e-test-utils-playwright/src/editor/index.ts

```diff
@@ -29,15 +35,20 @@ export class Editor {
 	browser: Browser;
 	page: Page;
 	context: BrowserContext;
+	useLegacyCanvas: boolean;
 
 	constructor( { page }: EditorConstructorProps ) {
 		this.page = page;
 		this.context = page.context();
 		this.browser = this.context.browser()!;
+		this.useLegacyCanvas = false;
 	}
 
-	get canvas(): Frame | Page {
-		return this.page.frame( 'editor-canvas' ) || this.page;
+	get canvas(): FrameLocator | Page {
+		if ( this.useLegacyCanvas ) {
+			return this.page;
+		}
```

Comment:
I wonder if we still need this. If the user has to opt-in to legacy canvas then they should be able to just use `page` than `editor.canvas`?

This also prevents the typescript awkwardness with `FrameLocator | Page` which doesn't really extend each other 😅 .

Reply:
This assumes that switching to legacy canvas is intentional, but that's not always true. Can we still return the `this.page` page as a fallback?

Reply:
I might be missing something, but won't `useLegacyCanvas` only be true if we call `switchToLegacyCanvas`? If the user opts in to call `switchToLegacyCanvas` then I guess we can expect them to use `page` instead.

Reply:
I don't like it as well and would be more than happy to remove it! 😄 There might be an issue with some Editor utils that reference canvas (e.g. `selectBlocks()` IIRC) while running against the legacy one, but it might as well just pass - will check it! 🤞

Reply:
Removed in https://github.com/WordPress/gutenberg/pull/54911/commits/fd96c75445906bc789f37c49c2e96fdeb90f36fb

Reply:
> It would be great to hide editors' implementation details for legacy canvas vs. iframe. It shouldn't be noticeable for people writing tests or using the editor, IMO.

This is also a good point. Unfortunately, I don't think automatically "piercing" through iframe boundaries can be done trivially in Playwright, yet (I don't know really, I haven't given it any thoughts 😅). There's a trade-off to make here, and I would personally prefer predictable types and autocomplete for package-level utility here.

Reply:
I missed a big part of this exchange - GH is not great at refreshing discussions. 😑 

Anyway, I've been on edge with this one because, on one side, there's the desirable autocomplete without the type ambiguity (`FrameLocator | Page`), and on the other side, there can be broken editor utils that depend on the `canvas` reference (like the aforementioned `selectBlocks()`. 

As for the solution, we can't do it the old "page fallback" way because `frameLocator` doesn't return `null` like the `frame` does – it will time out if not found. The only solution that worked for me (considering we want to hide the implementation detail) has been turning the `canvas` into an async getter and evaluating the parent element by wrapping framed and unframed locators with `Promise.any()`. Unfortunately, it would be still far from ideal as it also would:
- involve a hefty refactor caused by the inability to chain from `editor.canvas` (it would become `(await editor.canvas()).getByRole(...`.
- still mess with the canvas types.